### PR TITLE
#2898 Publish-Subsribe

### PR DESCRIPTION
**Pull request by Ehspresso**
This pull request solves issue #2898.

### CODE REVIEW

File: publisher-subscriber/README.md [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+            System.out.println("Invalid arguments. Should be: ");
+            System.out.println("java TLender [factory] [topic]");
+            System.exit(1);
+        }
+
+        try {
+            // Create and start activeMQ broker. Broker decouples publishers and subscribers.
+            //Additionally brokers manage threads and asynchronous sending and receiving of messages.
+            BrokerService broker = new BrokerService();
+            broker.addConnector("tcp://localhost:61616");
```
Comment (iluwatar) [inline publisher-subscriber/README.md:130]: should be a configuration parameter

File: publisher-subscriber/README.md [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+Press enter to quit application
+
+Running the class:
+
+The class must be run after the TLender class is running since TLender spins up the activeMQ broker.
+
+In order to see the messages being sent to multiple subscribers multiple instance of the TBorrower class need to be run. Either run multiple instances in an IDE or execute the following command in a command line from the root folder after generating the target folder:
+
+
+mvn exec:java -Dexec.mainClass=com.iluwatar.publishersubscriber.TBorrower -Dexec.args="TopicCF RateTopic 6"
```
Comment (iluwatar) [inline publisher-subscriber/README.md:196]: Let's move out these running instructions to a separate file and leave README.md purely for the pattern description. Maybe we can setup USAGE.md or something similar...

File: publisher-subscriber/README.md [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        return topic;
+    }
+
+    public double getCurrentRate() {
+        return currentRate;
+    }
+}
+```
+
+## Class diagram
```
Comment (iluwatar) [inline publisher-subscriber/README.md:318]: Class diagram is missing. Use urm-maven-plugin for that...

File: publisher-subscriber/pom.xml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        <dependency>
+            <groupId>org.apache.xbean</groupId>
+            <artifactId>xbean-spring</artifactId>
+            <version>4.24</version>
+        </dependency>
+        <dependency>
+            <groupId>log4j</groupId>
+            <artifactId>log4j</artifactId>
+            <version>1.2.17</version>
+        </dependency>
```
Comment (iluwatar) [inline publisher-subscriber/pom.xml:65]: We can use Lombok and slf4j for logging. No need for this one...

File: publisher-subscriber/src/test/java/com/iluwatar/publishersubscriber/TLenderTest.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  @Test
+  void testTLenderConstructor() {
+    //act
+    tLender = new TLender("TopicCF", "RateTopic");
+
+    //assert
+    Assertions.assertNotNull(tLender.gettConnection());
+    Assertions.assertNotNull(tLender.getTopic());
+    Assertions.assertNotNull(tLender.gettSession());
+  }
```
Comment (iluwatar) [inline publisher-subscriber/src/test/java/com/iluwatar/publishersubscriber/TLenderTest.java:32]: Check test coverage. I think this will not be sufficient.

**Review by iluwatar — CHANGES_REQUESTED** [review]

File: publisher-subscriber/src/main/java/com/iluwatar/publishersubscriber/TBorrower.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      //Retrieve request and response queues
+      topic = (Topic) context.lookup(topicName);
+
+      //Create subscriber and message listener
+      TopicSubscriber subscriber = tSession.createSubscriber(topic);
+      //Adds event listener to subscriber and uses onMessage as a callback
+      subscriber.setMessageListener(this);
+
+      tConnection.start();
+      System.out.println("Initial rate is " + currentRate + " \nWaiting for new rates...");
```
[the remark this answers is not in the document]
Comment (Ehspresso) [reply publisher-subscriber/src/main/java/com/iluwatar/publishersubscriber/TBorrower.java:51]: do you want all sout lines to be changed to logs?

File: publisher-subscriber/README.md (hunk 3 above)
Comment (Ehspresso) [reply publisher-subscriber/README.md]: I tried to get this to work, I can't seem to figure it out...perhaps this newest commit will show what I am doing wrong.

Comment (iluwatar) [reply publisher-subscriber/README.md]: The first step is to include your module in the parent pom.xml. After that, urm-maven-plugin should automatically process the module when doing maven build.

**Pull request closed without merging**